### PR TITLE
[BUGFIX] Fix `TF` shape computation for `pauli_decompose` 

### PR DESCRIPTION
**Pull request by obliviateandsurrender**
**Context:** There was an `if` condition inherited from the old legacy code where we were accessing shape via `H.shape`. This would fail for tensors built with TensorFlow. 

**Description of the Change:** This PR fixes it by using `qml.math.shape` instead.

**Benefits:** Differentiability on TensorFlow backend.

**Possible Drawbacks:** N/A

**Related GitHub Issues:** N/A

### CODE REVIEW

**Comment by timmysilv** [comment]
could you add a quick test for what was failing before? maybe.. do it for jax/torch as well just to be extra sure? otherwise this sounds good to me 👍

**Review by timmysilv — APPROVED** [review]
nice find!

File: pennylane/pauli/conversion.py [hunk 1]
```diff
@@ -221,12 +221,12 @@ def pauli_decompose(
...
 
     Args:
-        H (array[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`.
+        H (tensor_like[complex]): a Hermitian matrix of dimension :math:`2^n\times 2^n`.
         hide_identity (bool): does not include the Identity observable within
             the tensor products of the decomposition if ``True``.
         wire_order (list[Union[int, str]]): the ordered list of wires with respect
             to which the operator is represented as a matrix.
-        pauli (bool): return a PauliSentence instance if ``True``.
+        pauli (bool): return a ``PauliSentence`` instance if ``True``.
```
Comment (Jaybsoni) [inline pennylane/pauli/conversion.py:229]:
```suggestion
        pauli (bool): return a :class:`~.PauliSentence` instance if ``True``.
```

**Review by Jaybsoni — APPROVED** [review]
Nice 💯

**Pull request merged**